filebeat/input/journald/pkg/journalctl/reader.go

```diff
@@ -0,0 +1,272 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package journalctl
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os/exec"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/v7/filebeat/input/journald/pkg/journalfield"
+	input "github.com/elastic/beats/v7/filebeat/input/v2"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+// LocalSystemJournalID is the ID of the local system journal.
+const localSystemJournalID = "LOCAL_SYSTEM_JOURNAL"
+
+// ErrCancelled indicates the read was cancelled
+var ErrCancelled = errors.New("cancelled")
+
+// JournalEntry holds all fields of a journal entry plus cursor and timestamps
+type JournalEntry struct {
+	Fields             map[string]string
+	Cursor             string
+	RealtimeTimestamp  uint64
+	MonotonicTimestamp uint64
+}
+
+// Reader reads entries from journald by calling `jouranlctl`
+// and reading its output.
+type Reader struct {
```

Comment:
Did we ever add a comment explaining the issues forcing us to use `exec` instead of the systemd libraries? If not, we can just add a few lines and a link to the issue here.

Reply:
I added some reasoning there, let me know if it's enough. I did not mention the `exec` syscall, I kept the reasoning on a higher level of why using `journalctl`.